### DISCUSSION TITLE
Overriding Textbox.Text property

### DISCUSSION THREAD

**Discussion by Varmod** [AvaloniaUI/Avalonia#15528] [Q&A] [answered]
Hi,

I'd like to create a custom TextBox that overrides the TextBox.Text property and adds value coercion to it. 

I've tried the following, it compiles and doesn't complain but the `CoerceText` method is never called.

```c#
public static readonly StyledProperty<string?> TextProperty =
	TextBox.TextProperty.AddOwner<CustomTextBox>(
		new StyledPropertyMetadata<string?>(
			coerce: CoerceText, 
			defaultBindingMode: BindingMode.TwoWay));
```

What would be the best approach to do this?
Thank you

**Comment by Varmod** [comment]
Also on a side note, the [documentation](https://docs.avaloniaui.net/docs/guides/custom-controls/how-to-create-advanced-custom-controls#datavalidation-support) mentions that I should do this via a direct property but neither the `AddOwnerWithDataValidation` method is available to me and TextBox implements the property as a styled property and not direct. So I don't know how actual the information is.

**Comment by stevemonaco** [comment, marked as answer]
You shouldn't try to redefine dependency properties that are on a parent class. Instead, try using `TextProperty.OverrideMetadata`. `MaskedTextBox` uses it to override `PasswordChar` with coercion, but you should be able to adapt. https://github.com/AvaloniaUI/Avalonia/blob/76a1c70758884abcebc9e3665aa08290a841efa4/src/Avalonia.Controls/MaskedTextBox.cs#L47

TextBox is a complex control so I'm not sure if this will work as well as you intend. Coercion interrupts typing, so it might be bad UX over something like data validation, filtering individual keys, and showing error messages.

**Reply by Varmod** [reply]
Thank you, 
yes I was thinking about it wrongly. Coercing the textbox did indeed lead to poor UX. I decided to remake the functionality using data validation and it's better.
